kv/tikv/transaction/commands/command.go

```diff
@@ -0,0 +1,131 @@
+	PrepareWrites(txn *kvstore.MvccTxn) (interface{}, error)
+}
+
+// Run runs a transactional command.
+func RunCommand(cmd Command, innerServer inner_server.InnerServer, latches *latches.Latches) (interface{}, error) {
+	ctxt := cmd.Context()
+	var resp interface{}
+
+	keysToWrite := cmd.WillWrite()
+	if keysToWrite == nil {
+		// The command is readonly or requires access to the DB to determine the keys it will write.
+		reader, err := innerServer.Reader(ctxt)
+		if err != nil {
+			return nil, err
+		}
+		txn := kvstore.RoTxn{Reader: reader}
+		resp, keysToWrite, err = cmd.Read(&txn)
+		reader.Close()
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if keysToWrite != nil {
+		// The command will write to the DB.
+
+		latches.WaitForLatches(keysToWrite)
+		defer latches.ReleaseLatches(keysToWrite)
+
+		reader, err := innerServer.Reader(ctxt)
+		if err != nil {
+			return nil, err
+		}
+		defer reader.Close()
+
+		// Build an mvcc transaction.
+		txn := kvstore.NewTxn(reader)
+		resp, err = cmd.PrepareWrites(&txn)
+		if err != nil {
+			return nil, err
+		}
+
+		latches.Validate(&txn, keysToWrite)
+
+		// Building the transaction succeeded without conflict, write all writes to backing storage.
+		err = innerServer.Write(ctxt, txn.Writes)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return resp, nil
+}
+
+// CommandBase provides some default function implementations for the Command interface.
+type CommandBase struct {
+	context *kvrpcpb.Context
+}
+
+func (base CommandBase) Context() *kvrpcpb.Context {
+	return base.context
+}
+
+func (base CommandBase) Read(txn *kvstore.RoTxn) (interface{}, [][]byte, error) {
+	return nil, nil, nil
+}
+
+// ReadOnly is a helper type for commands which will never write anything to the database. It provides some default
+// function implementations.
+type ReadOnly struct{}
+
+func (ro ReadOnly) WillWrite() [][]byte {
+	return nil
+}
+
+func (ro ReadOnly) PrepareWrites(txn *kvstore.MvccTxn) (interface{}, error) {
+	return nil, nil
+}
+
+// regionError is a help method for handling region errors. If error is a region error, then it is added to resp (which
+// muse have a `RegionError` field; the response is returned. If the error is not a region error, then regionError returns
+// nil and the error.
+func regionError(err error, resp interface{}) (interface{}, error) {
+	if regionErr, ok := err.(*inner_server.RegionError); ok {
+		respValue := reflect.ValueOf(resp)
+		respValue.FieldByName("RegionError").Set(reflect.ValueOf(regionErr.RequestErr))
+		return resp, nil
+	}
+
+	return nil, err
+}
+
+// RawRegionError is similar to regionError but is specialized for the raw commands. They all have an `Error` field, so
+// if the error is not a region error, then it is set to that field in the response. This means that the error is always
+// handled and resp should be returned back to the gRPC layer.
+func RawRegionError(err error, resp interface{}) {
```

Comment:
let's move it to `server.go`, raw API should have nothing dependency on the transaction package.

Reply:
:+1: